### ISSUE TITLE
Mac BOINC client (7.16.14) is not considering the rsc_memory_bound when choosing to download and run a task

### ISSUE THREAD

**Issue by abowery** [BOINC/boinc#4200]
The Mac BOINC client (version 7.16.14) is not considering the rsc_memory_bound when choosing whether or not to download and run a task. The Mac BOINC client is downloading and running the task without checking whether it has sufficient RAM to run it. 

During the run of the task the Mac BOINC client is periodically checking the memory the task is using and if the memory the task is using is greater than the maximum memory set in the BOINC client configuration menu then the Mac BOINC client is quitting the task.

To verify this, I ran a task (the task had the rsc_memory_bound set to: 15000000000 (15GB)) in the Mac BOINC client. I turned on the 'mem_usage_debug' log flag. I got the following messages in the stdoutdae.txt file:

03-Feb-2021 15:21:54 [---]    max memory usage when active: 8192.00 MB
03-Feb-2021 15:21:54 [---]    max memory usage when idle: 14745.60 MB
.........
03-Feb-2021 20:59:34 [cpdnboinc_dev] [cpu_sched_debug] enforce: task openifs_e000_2015110112_90_d290_8842_0 can't run, too big 8419.28MB > 8192.00MB
03-Feb-2021 20:59:34 [cpdnboinc_dev] [mem_usage] job using too much memory, will preempt by quit

The 8192.00 MB figure is the figure I have set as the maximum RAM memory in the Mac BOINC client configuration. The 8419.28MB is the amount of RAM memory the task is using at that point. The Mac BOINC client is quitting the task because the RAM memory the task is using is greater than the RAM memory set in the Mac BOINC client configuration. The Mac BOINC client should not be downloading and running a task that requires 15GB RAM to run. It should be checking before it downloads whether it is able to run the task.

I think this behaviour may also apply to the rsc_disk_bound. I have not seen this behaviour with the Linux BOINC client.

**Steps To Reproduce**
1.  Turning on the 'mem_usage_debug' log flag on a Mac BOINC client. 
2. Press the 'update' button on the Mac BOINC client so that it downloads a task.
3. Monitor the file stdoutdae.txt to see the above messages.

**Expected behavior**
I expect the Mac BOINC client to check whether there is sufficient RAM available for the task before downloading and running the task.

**System Information**
 - OS:  Mac
 - BOINC Version: 7.16.14

**Comment by davidpanderson** [comment]
1) your preferences allow the use of 15,461,882,265 bytes of RAM when the computer is idle,
so that job can run then.  Does it?

2) the logic for deciding whether to send a job is in the server, not the client

**Comment by abowery** [comment]
@davidpanderson 
Ah okay, so it is downloading and running it properly. However when the machine is active I think it should be pausing the job and waiting for the machine to become idle again, rather than continuing the job and comparing the current memory usage, and then quitting the job when the memory usage goes above the 'max memory usage when active' value.

**Comment by abowery** [comment]
@davidpanderson 
Further update: In the computing preferences on the BOINC client I have the checkbox 'Suspend when computer not in use' unticked. This will explain the above behaviour. So I think the BOINC client is behaving correctly in this case.

One question I have is: here, when the BOINC client is telling the job to 'quit', I am getting it restart rather end. Is this what should happen in this case?

**Comment by davidpanderson** [comment]
I don't understand the last question.

**Comment by abowery** [comment]
@davidpanderson 
My question is: 
In this case described, when the RAM memory is exceeded the BOINC client chooses to restart the task. However if the disk usage limit is exceeded the BOINC client chooses to abort the task.
Shouldn't the BOINC client abort the task as well when the RAM memory is exceeded otherwise the task will be restarted repeatedly/infinitely?

**Comment by davidpanderson** [comment]
It should already do this.  app_control.cpp has:

    double max_ram = gstate.max_available_ram();
...
        if (atp->procinfo.working_set_size_smoothed > max_ram) {
            sprintf(buf, "working set size > client RAM limit: %.2fMB > %.2fMB",
                atp->procinfo.working_set_size_smoothed/MEGA, max_ram/MEGA
            );
            msg_printf(atp->result->project, MSG_INFO,
                "Aborting task %s: %s",
                atp->result->name, buf
            );
            atp->abort_task(EXIT_MEM_LIMIT_EXCEEDED, buf);
            did_anything = true;
            continue;
        }

**Comment by RichardHaselgrove** [comment]
Is that the right logic? The machine settings were
```
03-Feb-2021 15:21:54 [---]    max memory usage when active: 8192.00 MB
03-Feb-2021 15:21:54 [---]    max memory usage when idle: 14745.60 MB
```
The task was between those two limits

```03-Feb-2021 20:59:34 [cpdnboinc_dev] [cpu_sched_debug] enforce: task openifs_e000_2015110112_90_d290_8842_0 can't run, too big 8419.28MB > 8192.00MB```

So it was able to run when the machine was idle, but would be preempted and removed from memory when the machine was in use. As indeed it was:

```03-Feb-2021 20:59:34 [cpdnboinc_dev] [mem_usage] job using too much memory, will preempt by quit```

Presumably it then went into the state "suspended, waiting for memory", but we haven't seen that reported yet.

We are getting slightly muddled with the use of the generic term "quit". It would be better to use the formal BOINC terminology of "preempt" (a temporary state) and "abort" (final, irreversible) to distinguish the two actions.

**Comment by AenBleidd** [comment]
Was already fixed in #5669. Closing this as a duplicate
